File: resources/ios/FeedbackCenterState.swift
import Foundation

enum FeedbackCenterTone: String, CaseIterable {
    case `default`
    case success
    case warning
    case danger
}

enum FeedbackCenterWireEvent: Equatable {
    case press(callbackID: Int, nodeID: Int)
}

enum FeedbackCenterTiming {
    static let minimumDuration: TimeInterval = 4
    static let maximumMessageDuration: TimeInterval = 10
    static let actionExtension: TimeInterval = 2
    static let assistiveTechnologyMultiplier: Double = 2

    static func automaticDuration(
        message: String,
        hasAction: Bool,
        assistiveTechnologyActive: Bool
    ) -> TimeInterval {
        let additionalReadingTime = max(0, Double(message.count - 40) / 40)
        var duration = min(
            maximumMessageDuration,
            minimumDuration + additionalReadingTime
        )

        if hasAction {
            duration += actionExtension
        }
        if assistiveTechnologyActive {
            duration *= assistiveTechnologyMultiplier
        }

        return duration
    }
}

struct FeedbackCenterItemConfiguration: Equatable, Identifiable {
    let nodeID: Int
    let feedbackID: String
    let message: String
    let tone: FeedbackCenterTone
    let hold: Bool
    let actionLabel: String?
    let actionCallback: Int?
    let timeoutCallback: Int?
    let manualCallback: Int?

    private let hasValidIdentity: Bool
    private let hasValidMessage: Bool
    private let hasValidTone: Bool

    var id: String { feedbackID }

    var isEligible: Bool {
        hasValidIdentity
            && hasValidMessage
            && hasValidTone
            && (hold ? manualCallback != nil : timeoutCallback != nil)
    }

    init(node: NativeUINode) {
        nodeID = node.id

        let decodedID = node.props.getString("feedback_id")
        feedbackID = decodedID
        hasValidIdentity = !decodedID.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty

        let decodedMessage = node.props.getString("message")
        message = decodedMessage
        hasValidMessage = !decodedMessage.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty

        let decodedTone = FeedbackCenterTone(
            rawValue: node.props.getString("tone", default: "default")
        )
        tone = decodedTone ?? .default
        hasValidTone = decodedTone != nil

        hold = node.props.getBool("hold")

        let decodedActionLabel = node.props.getString("action_label")
            .trimmingCharacters(in: .whitespacesAndNewlines)
        let decodedActionCallback = Self.callback(node.props.getCallbackId("on_action"))
        if !decodedActionLabel.isEmpty, let decodedActionCallback {
            actionLabel = decodedActionLabel
            actionCallback = decodedActionCallback
        } else {
            actionLabel = nil
            actionCallback = nil
        }

        timeoutCallback = Self.callback(node.props.getCallbackId("on_timeout"))
        manualCallback = Self.callback(node.props.getCallbackId("on_manual"))
    }

    init(
        nodeID: Int,
        feedbackID: String,
        message: String,
        tone: FeedbackCenterTone = .default,
        hold: Bool = false,
        actionLabel: String? = nil,
        actionCallback: Int? = nil,
        timeoutCallback: Int? = 1,
        manualCallback: Int? = nil
    ) {
        self.nodeID = nodeID
        self.feedbackID = feedbackID
        self.message = message
        self.tone = tone
        self.hold = hold

        let trimmedAction = actionLabel?.trimmingCharacters(in: .whitespacesAndNewlines)
        if let trimmedAction, !trimmedAction.isEmpty, let actionCallback, actionCallback != 0 {
            self.actionLabel = trimmedAction
            self.actionCallback = actionCallback
        } else {
            self.actionLabel = nil
            self.actionCallback = nil
        }

        self.timeoutCallback = Self.callback(timeoutCallback ?? 0)
        self.manualCallback = Self.callback(manualCallback ?? 0)
        hasValidIdentity = !feedbackID.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
        hasValidMessage = !message.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
        hasValidTone = true
    }

    func automaticDuration(assistiveTechnologyActive: Bool) -> TimeInterval {
        FeedbackCenterTiming.automaticDuration(
            message: message,
            hasAction: actionCallback != nil,
            assistiveTechnologyActive: assistiveTechnologyActive
        )
    }

    private static func callback(_ value: Int) -> Int? {
        value == 0 ? nil : value
    }
}

struct FeedbackCenterQueueState {
    private(set) var items: [FeedbackCenterItemConfiguration] = []
    private(set) var elapsed: TimeInterval = 0
    private(set) var isPaused = false

    private var tombstones: Set<String> = []
    private var clock: TimeInterval
    private var assistiveTechnologyActive: Bool

    init(now: TimeInterval, assistiveTechnologyActive: Bool = false) {
        clock = now
        self.assistiveTechnologyActive = assistiveTechnologyActive
    }

    var visible: FeedbackCenterItemConfiguration? { items.first }
    var pendingIDs: [String] { items.dropFirst().map(\.feedbackID) }

    var visibleDuration: TimeInterval {
        guard let visible, !visible.hold else { return .infinity }
        return visible.automaticDuration(
            assistiveTechnologyActive: assistiveTechnologyActive
        )
    }

    var remaining: TimeInterval {
        max(0, visibleDuration - elapsed)
    }

    @discardableResult
    mutating func setAssistiveTechnologyActive(
        _ active: Bool,
        at now: TimeInterval
    ) -> FeedbackCenterWireEvent? {
        let expectedID = visible?.feedbackID
        synchronize(to: now)
        assistiveTechnologyActive = active
        return timeoutIfDue(feedbackID: expectedID, now: now)
    }

    @discardableResult
    mutating func reconcile(
        _ published: [FeedbackCenterItemConfiguration],
        now: TimeInterval
    ) -> FeedbackCenterWireEvent? {
        synchronize(to: now)

        let publishedIDs = Set(
            published.lazy
                .map(\.feedbackID)
                .filter { !$0.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty }
        )
        tombstones.formIntersection(publishedIDs)

        let idCounts = Dictionary(grouping: published, by: \.feedbackID).mapValues(\.count)
        let eligible = published.filter {
            $0.isEligible && idCounts[$0.feedbackID] == 1
        }
        let eligibleByID = Dictionary(uniqueKeysWithValues: eligible.map { ($0.feedbackID, $0) })

        let previousVisibleID = visible?.feedbackID
        var reconciled = items.compactMap { eligibleByID[$0.feedbackID] }
        let retainedIDs = Set(reconciled.map(\.feedbackID))

        for configuration in eligible
        where !retainedIDs.contains(configuration.feedbackID)
            && !tombstones.contains(configuration.feedbackID) {
            reconciled.append(configuration)
        }

        items = reconciled
        if visible?.feedbackID != previousVisibleID {
            elapsed = 0
            clock = now
        } else if visible == nil {
            elapsed = 0
            clock = now
        }

        return timeoutIfDue(feedbackID: previousVisibleID, now: now)
    }

    mutating func action(
        feedbackID: String,
        now: TimeInterval
    ) -> FeedbackCenterWireEvent? {
        guard visible?.feedbackID == feedbackID else { return nil }
        synchronize(to: now)
        guard let visible, let callbackID = visible.actionCallback else { return nil }
        return complete(
            feedbackID: visible.feedbackID,
            event: .press(callbackID: callbackID, nodeID: visible.nodeID),
            now: now
        )
    }

    mutating func timeout(
        feedbackID: String,
        now: TimeInterval
    ) -> FeedbackCenterWireEvent? {
        guard !isPaused, visible?.feedbackID == feedbackID else { return nil }
        synchronize(to: now)
        guard let visible, !visible.hold, let callbackID = visible.timeoutCallback else {
            return nil
        }
        return complete(
            feedbackID: visible.feedbackID,
            event: .press(callbackID: callbackID, nodeID: visible.nodeID),
            now: now
        )
    }

    mutating func manualDismiss(
        feedbackID: String,
        now: TimeInterval
    ) -> FeedbackCenterWireEvent? {
        guard visible?.feedbackID == feedbackID else { return nil }
        synchronize(to: now)
        guard let visible, visible.hold, let callbackID = visible.manualCallback else {
            return nil
        }
        return complete(
            feedbackID: visible.feedbackID,
            event: .press(callbackID: callbackID, nodeID: visible.nodeID),
            now: now
        )
    }

    @discardableResult
    mutating func advance(
        by interval: TimeInterval,
        feedbackID: String
    ) -> FeedbackCenterWireEvent? {
        guard interval > 0,
              !isPaused,
              let visible,
              visible.feedbackID == feedbackID,
              !visible.hold else { return nil }

        let advancedTime = clock + interval
        synchronize(to: advancedTime)
        return timeoutIfDue(feedbackID: feedbackID, now: advancedTime)
    }

    @discardableResult
    mutating func pause(at now: TimeInterval) -> FeedbackCenterWireEvent? {
        guard !isPaused else { return nil }
        let expectedID = visible?.feedbackID
        synchronize(to: now)
        let event = timeoutIfDue(feedbackID: expectedID, now: now)
        isPaused = true
        return event
    }

    @discardableResult
    mutating func resume(at now: TimeInterval) -> FeedbackCenterWireEvent? {
        guard isPaused else { return nil }
        clock = now
        isPaused = false
        return timeoutIfDue(feedbackID: visible?.feedbackID, now: now)
    }

    private mutating func synchronize(to now: TimeInterval) {
        defer { clock = max(clock, now) }
        guard now > clock, !isPaused, let visible, !visible.hold else { return }
        elapsed = min(visibleDuration, elapsed + (now - clock))
    }

    private mutating func complete(
        feedbackID: String,
        event: FeedbackCenterWireEvent,
        now: TimeInterval
    ) -> FeedbackCenterWireEvent? {
        guard visible?.feedbackID == feedbackID else { return nil }

        tombstones.insert(feedbackID)
        items.removeFirst()
        elapsed = 0
        clock = max(clock, now)
        return event
    }

    private mutating func timeoutIfDue(
        feedbackID: String?,
        now: TimeInterval
    ) -> FeedbackCenterWireEvent? {
        guard !isPaused,
              let feedbackID,
              visible?.feedbackID == feedbackID,
              visible?.hold == false,
              remaining == 0 else { return nil }

        return timeout(feedbackID: feedbackID, now: now)
    }
}

File: resources/android/FeedbackCenterState.kt
package dev.firstlightui.plugins.firstlight_ui.ui

import com.nativephp.mobile.ui.nativerender.NativeUINode
import kotlin.math.max
import kotlin.math.min

enum class FeedbackCenterTone(val wireName: String) {
    Default("default"),
    Success("success"),
    Warning("warning"),
    Danger("danger");

    companion object {
        fun fromWire(value: String): FeedbackCenterTone? = entries.firstOrNull { it.wireName == value }
    }
}

sealed interface FeedbackCenterWireEvent {
    data class Press(val callbackId: Int, val nodeId: Int) : FeedbackCenterWireEvent
}

object FeedbackCenterTiming {
    const val MinimumDurationMillis = 4_000L
    const val MaximumMessageDurationMillis = 10_000L
    const val ActionExtensionMillis = 2_000L

    fun automaticBaseMillis(message: String, hasAction: Boolean): Long {
        val additionalReadingMillis = max(0, message.length - 40).toLong() * 1_000L / 40L
        val messageDuration = min(MaximumMessageDurationMillis, MinimumDurationMillis + additionalReadingMillis)
        return messageDuration + if (hasAction) ActionExtensionMillis else 0L
    }
}

data class FeedbackCenterItemConfiguration(
    val nodeId: Int,
    val feedbackId: String,
    val message: String,
    val tone: FeedbackCenterTone,
    val hold: Boolean,
    val actionLabel: String?,
    val actionCallback: Int?,
    val timeoutCallback: Int?,
    val manualCallback: Int?,
    private val hasValidIdentity: Boolean,
    private val hasValidMessage: Boolean,
    private val hasValidTone: Boolean,
) {
    constructor(node: NativeUINode) : this(
        nodeId = node.id,
        feedbackId = node.props.getString("feedback_id"),
        message = node.props.getString("message"),
        tone = FeedbackCenterTone.fromWire(node.props.getString("tone", "default")) ?: FeedbackCenterTone.Default,
        hold = node.props.getBool("hold"),
        actionLabel = validActionLabel(node),
        actionCallback = validActionCallback(node),
        timeoutCallback = callback(node.props.getCallbackId("on_timeout")),
        manualCallback = callback(node.props.getCallbackId("on_manual")),
        hasValidIdentity = node.props.getString("feedback_id").isNotBlank(),
        hasValidMessage = node.props.getString("message").isNotBlank(),
        hasValidTone = FeedbackCenterTone.fromWire(node.props.getString("tone", "default")) != null,
    )

    val hasAction: Boolean get() = actionLabel != null && actionCallback != null

    val isEligible: Boolean
        get() = hasValidIdentity && hasValidMessage && hasValidTone &&
            manualCallback != null && (hold || timeoutCallback != null)

    companion object {
        private fun callback(value: Int): Int? = value.takeIf { it != 0 }

        private fun validActionLabel(node: NativeUINode): String? {
            val label = node.props.getString("action_label").trim()
            return label.takeIf { it.isNotEmpty() && validActionCallback(node) != null }
        }

        private fun validActionCallback(node: NativeUINode): Int? {
            val label = node.props.getString("action_label").trim()
            return callback(node.props.getCallbackId("on_action")).takeIf { label.isNotEmpty() }
        }
    }
}

class FeedbackCenterQueueState(
    nowMillis: Long,
    timeoutPolicy: (FeedbackCenterItemConfiguration) -> Long = {
        FeedbackCenterTiming.automaticBaseMillis(it.message, it.hasAction)
    },
) {
    private val items = mutableListOf<FeedbackCenterItemConfiguration>()
    private val tombstones = mutableSetOf<String>()
    private var clockMillis = nowMillis
    private var timeoutPolicy = timeoutPolicy

    var elapsedMillis: Long = 0
        private set
    var isPaused: Boolean = false
        private set

    val visible: FeedbackCenterItemConfiguration? get() = items.firstOrNull()
    val pendingIds: List<String> get() = items.drop(1).map { it.feedbackId }
    val visibleDurationMillis: Long
        get() = visible?.takeUnless { it.hold }?.let(timeoutPolicy) ?: Long.MAX_VALUE
    val remainingMillis: Long get() = max(0L, visibleDurationMillis - elapsedMillis)

    fun setTimingPolicy(
        policy: (FeedbackCenterItemConfiguration) -> Long,
        nowMillis: Long,
    ): FeedbackCenterWireEvent? {
        val expectedId = visible?.feedbackId
        synchronize(nowMillis)
        timeoutPolicy = policy
        return timeoutIfDue(expectedId, nowMillis)
    }

    fun timingPolicyChanged(nowMillis: Long): FeedbackCenterWireEvent? {
        val expectedId = visible?.feedbackId
        synchronize(nowMillis)
        return timeoutIfDue(expectedId, nowMillis)
    }

    fun reconcile(
        published: List<FeedbackCenterItemConfiguration>,
        nowMillis: Long,
    ): FeedbackCenterWireEvent? {
        synchronize(nowMillis)

        val publishedIds = published.map { it.feedbackId }.filter { it.isNotBlank() }.toSet()
        tombstones.retainAll(publishedIds)

        val idCounts = published.groupingBy { it.feedbackId }.eachCount()
        val eligible = published.filter { it.isEligible && idCounts[it.feedbackId] == 1 }
        val eligibleById = eligible.associateBy { it.feedbackId }
        val previousVisibleId = visible?.feedbackId

        val reconciled = items.mapNotNull { eligibleById[it.feedbackId] }.toMutableList()
        val retainedIds = reconciled.mapTo(mutableSetOf()) { it.feedbackId }
        eligible.forEach { configuration ->
            if (retainedIds.add(configuration.feedbackId) && configuration.feedbackId !in tombstones) {
                reconciled += configuration
            }
        }

        items.clear()
        items += reconciled
        if (visible?.feedbackId != previousVisibleId || visible == null) {
            elapsedMillis = 0
            clockMillis = nowMillis
        }

        return timeoutIfDue(previousVisibleId, nowMillis)
    }

    fun action(feedbackId: String, nowMillis: Long): FeedbackCenterWireEvent? {
        if (visible?.feedbackId != feedbackId) return null
        synchronize(nowMillis)
        val current = visible ?: return null
        val callback = current.actionCallback ?: return null
        return complete(current.feedbackId, FeedbackCenterWireEvent.Press(callback, current.nodeId), nowMillis)
    }

    fun timeout(feedbackId: String, nowMillis: Long): FeedbackCenterWireEvent? {
        if (isPaused || visible?.feedbackId != feedbackId) return null
        synchronize(nowMillis)
        val current = visible?.takeUnless { it.hold } ?: return null
        val callback = current.timeoutCallback ?: return null
        return complete(current.feedbackId, FeedbackCenterWireEvent.Press(callback, current.nodeId), nowMillis)
    }

    fun manualDismiss(feedbackId: String, nowMillis: Long): FeedbackCenterWireEvent? {
        if (visible?.feedbackId != feedbackId) return null
        synchronize(nowMillis)
        val current = visible ?: return null
        val callback = current.manualCallback ?: return null
        return complete(current.feedbackId, FeedbackCenterWireEvent.Press(callback, current.nodeId), nowMillis)
    }

    fun advanceBy(intervalMillis: Long, feedbackId: String): FeedbackCenterWireEvent? {
        val current = visible
        if (intervalMillis <= 0 || isPaused || current?.feedbackId != feedbackId || current.hold) return null
        val advancedTime = clockMillis + intervalMillis
        synchronize(advancedTime)
        return timeoutIfDue(feedbackId, advancedTime)
    }

    fun pause(nowMillis: Long): FeedbackCenterWireEvent? {
        if (isPaused) return null
        val expectedId = visible?.feedbackId
        synchronize(nowMillis)
        val event = timeoutIfDue(expectedId, nowMillis)
        isPaused = true
        return event
    }

    fun resume(nowMillis: Long): FeedbackCenterWireEvent? {
        if (!isPaused) return null
        clockMillis = nowMillis
        isPaused = false
        return timeoutIfDue(visible?.feedbackId, nowMillis)
    }

    private fun synchronize(nowMillis: Long) {
        try {
            val current = visible
            if (nowMillis > clockMillis && !isPaused && current != null && !current.hold) {
                elapsedMillis = min(visibleDurationMillis, elapsedMillis + (nowMillis - clockMillis))
            }
        } finally {
            clockMillis = max(clockMillis, nowMillis)
        }
    }

    private fun complete(
        feedbackId: String,
        event: FeedbackCenterWireEvent,
        nowMillis: Long,
    ): FeedbackCenterWireEvent? {
        if (visible?.feedbackId != feedbackId) return null
        tombstones += feedbackId
        items.removeAt(0)
        elapsedMillis = 0
        clockMillis = max(clockMillis, nowMillis)
        return event
    }

    private fun timeoutIfDue(feedbackId: String?, nowMillis: Long): FeedbackCenterWireEvent? {
        if (
            isPaused || feedbackId == null || visible?.feedbackId != feedbackId ||
            visible?.hold != false || remainingMillis != 0L
        ) return null
        return timeout(feedbackId, nowMillis)
    }
}
